### PR TITLE
Snap gate tutorial with qgrad

### PR DESCRIPTION
**Pull request by araza6**
Implementation of [SNAP gates paper](https://arxiv.org/abs/2004.14256) using qgrad. The qgrad methods work fine, just that the gradient descent does not seem to converge. I have spent some time in discerning what the reason might be, so may be another pair of eyes can help to suggest what is going wrong here.

### CODE REVIEW

**Comment by araza6** [comment]
Also, the `Displace` function had to be changed here in `qgrad_qutip` since it was not fully compatible with JAX earlier on

**Comment by nathanshammah** [comment]
We set up the `qgrad-examples` repo for tutorials. @araza6 can you reopen this PR there?

**Comment by araza6** [comment]
I found this while testing my notebook. So when I apply the blocks to the vacuum state, the norm of the evolved kets seems to be progressively changing. I have the following function `add_blocks` that applies the *whole* set of blocks on the vacuum state.

```

displace = Displace(N)  # initialize Displace object 

def add_blocks(initial, T,  hilbert_size, params):
    ""Applies T blocks of operators to the initial state, where each of the T blocks is
      a collection of operators, 

        :: math : `D(\alpha) SNAP(\theta) D(-\alpha),
   
      where :math:`D(\alpha)` is the displacement operator with displacement :math:`alpha` and :math:`SNAP(\theta)`
      is the SNAP gate with parameter vector :math:`\theta` of length N, the size of the Hilbert space.

    Args:
    ------
        initial (jnp.ndarray): initial state to apply blocks on (ket |0> in our case)
        T (int): number of blocks to apply
        hilbert_size (int): Size of the Hilbert space
        params (jnp.ndarray): parameter array of alphas and that's of size :math: `T * hilbert_size + T`, wherein the first 
                     T parameters are alphas and the rest are T hilbert_size-dimensional vectors representing corresponding theta vectors.

    Returns:
    -----------
         evolved (jnp.array): hilbert_size * 1-dimensional array representing the action of the blocks on the vacuum state
   
    for t in range(T):
        blk = jnp.dot(displace(params[0 + t]), initial)
        blk = jnp.dot(snap(hilbert_size, params[T + (t * hilbert_size) : T + (t + 1) * hilbert_size]), blk)
        initial = jnp.dot(displace(- params[0 + t]), blk) # displace(alpha)^{\dagger} = displace(-alpha)

    evolved = initial
    return evolved
```

When I apply this on the input ket, |0>, the norm decreases with the application of each of the displace and SNAP operations *within* any single block (and of course after the application of the T blocks). Take T = 3 for example with the following initial vacuum state and target binomial state:

```
N = 10
T = 3
initial = basis(N, 0)
target = (jnp.sqrt(3) * basis(10, 3) +  basis(10, 9)) / 2.0
```

Now when I calculate the norm of the state I get after the application of these 3 blocks on the initial state, the norm of the quantum state is not preserved since `jnp.linalg.norm(add_blocks(initial, params,T, N))` does not return 1. I am wondering why would that be? If the transformation is unitary then the norm should be preserved.

**Comment by quantshah** [comment]
Could you check this in QuTiP? There is sometimes an issue with setting the Hilbert space cutoff. In the paper they choose 100.

**Comment by araza6** [comment]
Hilbert space cutoff in the paper 10, I suppose

**Comment by araza6** [comment]
A bit more debugging reveals that the error is in the *third* application of a block. For T = 3, I print norms after the application of each of the blocks 

```
initial = basis(N, 0)
target = (jnp.sqrt(3) * basis(10, 3) +  basis(10, 9)) / 2.0
params = jnp.concatenate((alphas, theta1, theta2, theta3)).reshape(3 * N + 3, 1)

#block1 
tmp = jnp.dot(displace(params[0]), initial)
tmp = jnp.dot(snap(N, params[3:3+N]), tmp)
tmp = jnp.dot(displace(-params[0]), tmp)
print(jnp.linalg.norm(tmp))

# block 2
tmp = jnp.dot(displace(params[1]), tmp)
tmp = jnp.dot(snap(N, params[3+N:3+2*N]), tmp)
tmp = jnp.dot(displace(-params[1]), tmp)
print(jnp.linalg.norm(tmp))

# block 3
tmp = jnp.dot(displace(params[2]), tmp)
print(jnp.linalg.norm(tmp))
tmp = jnp.dot(snap(N, params[3+2*N:3+3*N]), tmp)
print(jnp.linalg.norm(tmp))
tmp = jnp.dot(displace(-params[2]), tmp)

print(jnp.linalg.norm(tmp))
```
And the norm changes right after applying SNAP in from the third block as can be seen in the following norm values

0.9999991
0.99999815
0.9999977
2.8582945
2.8582935

The same error persists when I take T = 5. At the snap gate of the *third* block, norm changes to 2.8582935, but after that the norm is preserved

**Comment by quantshah** [comment]
I see. Could you show the parameters of the third block?

**Comment by quantshah** [comment]
The snap gate parameters should be real valued thetas I think such that the phase added is "exp(-1j*theta)|n>" , [see the original paper](https://arxiv.org/pdf/1503.01496.pdf). If for some reason your theta values are complex (a + 1j b) then you will add an extra factor exp(b) to your Fock states. Careful about that.

**Comment by quantshah** [comment]
Whenever you post some code, make sure you send a working script with the parameters (if it is too long then use something like https://paste.ubuntu.com) so that others can reproduce the errors easily. In most cases you can make a short snippet (with the imports) so that others can copy-paste the code and run it to check it out. Helps with the debugging.

**Comment by araza6** [comment]
The tutorial is now complete from my side and ready for review (had so much fun putting this to work!!). 

Here is a caveat though. The gradient descent routine takes a lot of time, perhaps because of the auto diff evaluations. But that is just JAX's implementation. I am not sure what we can do to bolster the run time. That, however, is not a huge concern, I believe since the aim is to show how to work with qgrad.

In addition to the .py file, I have left the .ipynb notebook for ease of viewing Wigner functions and hinton plot.

**Comment by quantshah** [comment]
```
import numpy as np
from qutip import coherent
from qutip.visualization import plot_wigner


def gkp(hilbert_size, delta=None, mu = None):
    """Generates a GKP state
    """
    gkp = 0*coherent(hilbert_size, 0)

    c = np.sqrt(np.pi/2)

    if mu is None:
        mu = np.random.randint(2)

    if delta is None:
        delta = np.random.uniform(0.01, .99)

    zrange = range(-20, 20)

    for n1 in zrange:
        for n2 in zrange:        
            a = c*(2*n1 + mu + 1j*n2)
            alpha = coherent(hilbert_size, a)
            gkp += np.exp(-delta**2*np.abs(a)**2)*np.exp(-1j*c**2 * 2*n1 * n2)*alpha

    rho = gkp*gkp.dag()
    return rho.unit()


rho = gkp(32, 0.2)
plot_wigner(rho)
```

File: examples/Efficient Cavity Control with SNAP Gates.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+alphas = jnp.array([1., 0.5, 1.]) # Displace parameters
+theta1, theta2, theta3 = [0.5], [0.5, 1.5, 0.5], [0.5, 1.5, 0.5, 1.3] # SNAP parameters
+# NOTE: No input values to JAX differentiable functions should be int
+
+theta1, theta2, theta3 = pad_thetas(N, theta1), pad_thetas(N, theta2), pad_thetas(N, theta3)
+
+initial = basis(N, 0) # initial vacuum state
+target = (jnp.sqrt(3) * basis(10, 3) +  basis(10, 9)) / 2.0 # target state b1 as shown above|
+
```
Comment (quantshah) [inline examples/Efficient Cavity Control with SNAP Gates.py:148]: This block can be removed since we already have an initialisation before.

File: examples/Efficient Cavity Control with SNAP Gates.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        prev_params = params
+        der = der_cost(prev_params, initial, target, T, N)
+        params = prev_params - lr * der 
+        iters += 1
+    cost_val = cost(params, initial, target, T, N)
+    progress = [epoch+1, cost_val]
+    if ((epoch + 2) % 2 != 0):
+        params_hist.append(params)
+        print("Epoch: {:2f} | Cost: {:3f}".format(*jnp.asarray(progress)))
+
```
Comment (quantshah) [inline examples/Efficient Cavity Control with SNAP Gates.py:179]: The sub iterations iters < max_iters is confusion. We discussed this before too that what this does is simply do epochs * max_iters iteration. You can remove this and just increase the number of epochs as


```
params_hist = [(alphas, thetas),]
epochs = 100
lr = .1 # learning rate

for epoch in range (epochs):
    prev_thetas = thetas
    prev_alphas = alphas
    
    der = der_cost(prev_alphas, prev_thetas, initial_state, target_state)
    
    alphas = prev_alphas - lr * der[0]
    thetas = prev_thetas - lr * der[1]

    cost_val = cost(alphas, thetas, initial_state, target_state)
    progress = [epoch+1, cost_val]

    if ((epoch) % 10 != 0):
        params_hist.append((alphas, thetas))
        print("Epoch: {:2f} | Cost: {:3f}".format(*jnp.asarray(progress)))
```

File: examples/Efficient Cavity Control with SNAP Gates.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+# In[13]:
+
+
+def display_evolution(parameters):
```
Comment (quantshah) [inline examples/Efficient Cavity Control with SNAP Gates.py:184]: With the new show_state function this can be simplified. I would maybe suggest instead of looping through it all we could do something like for 1000 epochs if the user is asking for 4 plots then we show the plots at step 0, 250, 500, 750 and the end. Try to change the logic to that using the function below. 

```
def display_evolution(parameters, num_plots = 4):
    """
    Displays the intermediate states during the learning schedule
    """
    show_state(initial)

    for i in range(0, num_plots):
        alphas, thetas = parameters[i]
        x = apply_blocks(alphas, thetas, initial)
        show_state(x)
    
    # final parameters after last epoch
    alphas, thetas = parameters[-1]
    x = apply_blocks(alphas, thetas, initial)
    show_state(x)
    
display_evolution(params_hist)
```

File: examples/Efficient Cavity Control with SNAP Gates.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    
+display_evolution(params_hist)
+
+
+# # Comments on Learning
+# We see that starting from a vacuum state $|0>$, we efficiently learn the target state
+# $b_{1} = \frac{\sqrt3 |3> + |9>}{2}$. The desired target state's Hinton plot and Wigner function are shown below. It can be seen that the last row above is almost the same as the target state shown below, implying we learn parameters $\alpha$ and $\vec \theta$ efficiently so as to go from the vacuum state to a desired target state (in this case a binomial state) in just three applications of blocks, $\hat B$. This tutorial is just meant to show the usage of `qgrad`. One may increase the number of epochs to achieve even better learning.
+
+# In[15]:
+
```
Comment (quantshah) [inline examples/Efficient Cavity Control with SNAP Gates.py:216]: This could be titled as `Conclusion`

Note that we have not used the smart block initialisation that the Fossel paper uses, mention this in the beginning and also here so that users are not confused since we are not doing everything that the Fossel paper does. 

Also another point to mention would be that we use a very simple gradient update rule here, a simple Euler forward method. This could be replaced for instance with Adam: https://jax.readthedocs.io/en/latest/_modules/jax/experimental/optimizers.html#adam 

This is very easy to implement with Jax: 

Note that for 1000 iterations I got stuck in some local minima with the following loss and final state

```
from jax.experimental import optimizers
from jax import jit


alphas = jnp.array([1., 0.5, 1.]) # Displace parameters

theta1, theta2, theta3 = [0.5], [0.5, 1.5, 0.5], [0.5, 1.5, 0.5, 1.3] # SNAP parameters
# NOTE: No input values to JAX differentiable functions should be int
thetas = jnp.array([pad_thetas(N, p) for p in [theta1, theta2, theta3]])

opt_init, opt_update, opt_get_params = optimizers.adam(step_size=0.01)
opt_state = opt_init((alphas, thetas))

# Define a compiled update step

@jit
def step(i, opt_state, initial, target):
    alphas, thetas = opt_get_params(opt_state)
    g = der_cost(alphas, thetas, initial_state, target_state)
    return opt_update(i, g, opt_state)


loss = []

for i in range(1000):
    opt_state = step(i, opt_state, initial_state, target_state)
    alphas, thetas = opt_get_params(opt_state)
    loss.append(cost(alphas, thetas, initial_state, target_state))

    if i % 200 == 0:        
        state_evolved = apply_blocks(alphas, thetas, initial_state)
        show_state(state_evolved)
        
plt.plot(loss)
```


Unknown-2


image

Comment (quantshah) [reply examples/Efficient Cavity Control with SNAP Gates.py:216]: Note that this might be due to jax optimisers not working with complex numbers or the way we have separated alphas and thetas in the params and not really let them be some matrix or vector.

I do get some warning there for complex numbers. So please try to check that. Alternatively it is extremely straight forward just to replace your gradient descent lines with the following from jax optimisers which implements Adam. Here we will need to initialise our own m0, v0 for the alphas and thetas and update both of them (not just x here which is probably just considered to be the full vector of parameters)

```
  step_size = make_schedule(step_size)
  def init(x0):
    m0 = jnp.zeros_like(x0)
    v0 = jnp.zeros_like(x0)
    return x0, m0, v0
  def update(i, g, state):
    x, m, v = state
    m = (1 - b1) * g + b1 * m  # First  moment estimate.
    v = (1 - b2) * jnp.square(g) + b2 * v  # Second moment estimate.
    mhat = m / (1 - b1 ** (i + 1))  # Bias correction.
    vhat = v / (1 - b2 ** (i + 1))
    x = x - step_size(i) * mhat / (jnp.sqrt(vhat) + eps)
    return x, m, v
  def get_params(state):
    x, m, v = state
    return x
  return init, update, get_params
```

**Comment by nathanshammah** [comment]
I remove my review and will let @quantshah review/approve for merge once the PR is ready.

File: examples/Efficient Cavity Control with SNAP Gates.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# # Introduction
+# This tutorial reproduces [Thomas et al](https://arxiv.org/abs/2004.14256), titled **Efficient cavity control with SNAP gates**. Here the aim is to apply a seqeunce of Displace operators and SNAP gates, called blocks, $\hat{B}$ on an initial 10-dimensional vacuum state, $|0>$ with just the right parameters such that with the action of three such blocks, $\hat{B}$ we land on to the target binomial state $b_{1}$
```
[the remark this answers is not in the document]
Comment (araza6) [reply examples/Efficient Cavity Control with SNAP Gates.py:5]: Thanks for reviewing. Why do you add double backslash for say `\dagger` instead of `\\dagger`?

**Comment by quantshah** [comment]
It's just that Sphinx treats the first slash as escape.

**Comment by araza6** [comment]
@quantshah I have inculcated your PR review comments and snap notebook in this cleaner version of the tutorial. You may re-review it and let me know if it needs further work.

File: examples/Efficient Cavity Control with SNAP Gates.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+    opt_state = step(i, opt_state, opt_update)
+    params = get_params(opt_state)
+    params_hist.append(params)
+    f = 1 - cost(params, initial_state, target_state)
+    loss_val.append(1 - f)
+    pbar.set_description("Fidelity {}".format(f))
+
+
+
+def display_evolution(parameters, num_plots = 4):
```
Comment (quantshah) [inline examples/Efficient Cavity Control with SNAP Gates.py:289]: This function should document the variables parameters and num_plots

File: examples/Efficient Cavity Control with SNAP Gates.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+# of blocks, $\hat B$. This tutorial is just meant to show the 
+# usage of `qgrad`. Like the authors in the original paper, one
+# may use more sophistacated techniques like gradient clipping,
+# regularized cost function, and greater number of epochs to
+# achieve even better learning.
+# 
+
+
+
+plt.plot(loss_val)
```
Comment (quantshah) [inline examples/Efficient Cavity Control with SNAP Gates.py:331]: You could show the loss and the fidelity together in this same plot. Physics folks would like to see the fidelity. You can plot both in the same plot and use a legend to specify which is loss and which is fidelity. Alternatively just plot fidelity.

**Comment by quantshah** [comment]
@araza6 Very nicely done. I have only two small comments. Please address them, update the ipynb notebook and feel free to merge this.

**Comment by quantshah** [comment]
@araza6 Make sure you pull the latest master into this branch (or rebase against the master) before merging. The same applies for the other PRs.
